lib/Raven/Client.php

```diff
@@ -233,35 +238,49 @@ class Client
      */
     protected $_shutdown_function_has_been_set;
 
-    public function __construct($options_or_dsn = null, $options = [])
+    /**
+     * @var Configuration The client configuration
+     */
+    protected $config;
+
+    /**
+     * Constructor.
+     *
+     * @param Configuration $config The client configuration
+     */
+    public function __construct(Configuration $config)
```

Comment:
Yes, I could create inside the constructor, but this way we can

- Have a builder class that creates the client, given the amount of options we have
- We **make explicit** that the `Client` class **depends** from the `Configuration` class
- We could have a static method (something like `createClient`) that will accept the options as an array and just call the constructor
- Users can extend the `Configuration` class if they want without the need to subclass the `Client` class, or they can make changes to the configuration before instantiating the client by passing the configuration to some code outside our control (arrays are passed by value, objects not)

Plus:

- The Ruby SDK is done in the exact same way, and noone complains
- There are a lot of pros, the only cons until now is that the programmer must add a new line to create the configuration object before instantiating the client. Not so hard, huh?